### ISSUE TITLE
Super Monaco GP - audio inconsistencies

### ISSUE THREAD

**Issue by nickcsbr** [MiSTer-devel/Genesis_MiSTer#173]
Hello,

I have noticed audio inconsistencies at the Super Monaco GP game. The engine does not sound accurate to the real hardware and it varies from race to race. The other cars' engine sound also seem to suffer from the same behaviour. I cannot tell for sure but I believe those sounds are generated by the noise channel. Please let me know if you need any other information regarding this issue. Thank you and keep up this amazing work.

**Comment by jotego** [comment]
Interesting. By the way, there is no noise channel in the YM2612. I wonder what technique the CPU uses for this.

**Comment by birdybro** [comment]
> 
> 
> Interesting. By the way, there is no noise channel in the YM2612. I wonder what technique the CPU uses for this.

The SN76489AN (DCSG) was what had the noise channel in the genesis, and this core uses your jt89 core as a result.

https://github.com/jotego/jt89

https://github.com/MiSTer-devel/Genesis_MiSTer/tree/master/rtl/jt89

However, does the jt89 core account for the differences between the discrete one used in arcade systems and the integrated one used in the Genesis? In the Genesis the SN76489 DCSG was integrated into the VDP and had some differences:

```
Sega used real SN76489AN chips in their SG-1000 game console and SC-3000 computer, but used SN76489A clones in their Master System, Game Gear, and Sega Genesis game consoles. These modified sound chips were incorporated into the systems' video display processor (VDP). Although basic functionality is almost identical to that of the original SN76489A sound processor, a few small differences existed: the randomness for the noise channel is generated differently, and the Game Gear's version includes an additional flag register that designates which speaker(s) each audio channel are output (left, right, or both). The periodic noise is also 16 stages long on the Sega-made clones rather than 15; this makes a significant difference for music/programs which use periodic noise, as sounds will play at 6.25% lower pitch than on the TI-made chips.
```

https://en.wikipedia.org/wiki/Texas_Instruments_SN76489#Clones_and_successors

**Comment by birdybro** [comment]
Looking into this a little more. JT89 has been updated by jotego and the genesis core doesn't reflect these changes.

From jt89_noise.v
![image](https://user-images.githubusercontent.com/16388068/109842139-ff4fcd00-7c06-11eb-9432-12c8e22e00f7.png)

From jt89_mixer.v
![image](https://user-images.githubusercontent.com/16388068/109842216-14c4f700-7c07-11eb-9586-4a46737f7f80.png)

from jt89.v
![image](https://user-images.githubusercontent.com/16388068/109842255-1f7f8c00-7c07-11eb-9fb2-be7369fccb5b.png)
![image](https://user-images.githubusercontent.com/16388068/109842360-3b832d80-7c07-11eb-8458-763735603e6a.png)

Would this have an impact on this issue @jotego ?

**Comment by nickcsbr** [comment]
> Interesting. By the way, there is no noise channel in the YM2612. I wonder what technique the CPU uses for this.

Thank you for pointing out that there is no noise channel in the YM2612. I was misled by the Mega SG's audio settings (which also suffers to accurately replicate the engine sounds in this particular game, but in a different way).

**Comment by birdybro** [comment]
I'm compiling a test core which updates the jt89 core used in he Genesis core matched to jotego's current implementation.

I'll give you a test file in a bit so you can try it out and see if the problem goes away.

**Comment by birdybro** [comment]
[Genesisjt89test.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/6078151/Genesisjt89test.zip)

Try that out for me and see. It may just not work at all. Not an expert on this.

**Comment by nickcsbr** [comment]
> [Genesisjt89test.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/6078151/Genesisjt89test.zip)
> 
> Try that out for me and see. It may just not work at all. Not an expert on this.

Thanks! I'll give it a try and let you know what happens.

**Comment by birdybro** [comment]
That might not work actually. so doing you another one in case it doesn't with a slight change.

**Comment by nickcsbr** [comment]
The file did work perfectly but the sound issue is still there. In some races the sound was spot on, but on the others, it is off. Maybe worth mentioning what I am doing: I start a World Championship, then select 7 speed transmission and start a race.

**Comment by birdybro** [comment]
[Genesisjt89test2.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/6078354/Genesisjt89test2.zip)

One more time.

If it's still off, I'll do og hardware comparisons recorded to maybe point us to what's going on. :)

**Comment by nickcsbr** [comment]
Still no luck. Thanks for trying!

**Comment by jotego** [comment]
I don’t remember well the details of this chip, other than it was small and easy to do but with contradicting information about it on the internet. 
I will follow your comments in case there’s something to change.

**Comment by birdybro** [comment]
This sound effect doesn't get picked up by VGM logging, which is somewhat interesting to me. I'm wondering if this is a sample that is being passed through the PSG, which is a capability it had but was very rarely used.

**Comment by birdybro** [comment]
Audio comparisons here in case anyone wants to know what the OP is talking about:

[engine-md1va3.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/6080660/engine-md1va3.zip) - Recorded on my MD1-VA3 genesis.

[engine-mister-20210225.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/6080663/engine-mister-20210225.zip)
 - Recorded on the mister core (released 02-25-2021)

**Comment by birdybro** [comment]
More comparisons since it might be differences with jt12 afterall... Using a folder comparison between the two and just noting the things that stand out in case it helps.

* clk_en_2 missing from mister core in multiple files
* enable_irq_a/b are swapped in the mister core comparably in jt12_top.v
* acmd_up_b isn't being used by the mister core
* in jt12_eg_ctrl.v line 102, mister core is using `>=` instead of `==`
* jt12_div.v - mister core has num_ch=6 added (or maybe that was present in old version).
* through multiple files the `ladder` input is present in the mister core, but not present elsewhere. Maybe this is a framework necessity I'm unfamiliar with.
* /mixer/jt12_genmix.v line 78 posedge from jotego's core switched for negedge (probably good reason for that).

**Comment by birdybro** [comment]
[engine_MiSTer_Unstable_20220423.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/8548112/engine_MiSTer_Unstable_20220423.zip)

As of latest unstable build this morning, this seems to be resolved.

Referencing jt89 update from Sorgelig --> https://github.com/MiSTer-devel/Genesis_MiSTer/issues/209

**Comment by nickcsbr** [comment]
I tested it with the latest unstable build but unfortunately the problem still persists. It seems to be a random issue as it may happen immediately at first or after some races.

**Comment by davedoc1984** [comment]
Engine Note still sounds incorrect compared to Real Hardware. Almost as if the pitch of it is too high?

I can try and capture some comparison Audio of both the Core and a Japanese Model 1 Mega Drive to demonstrate?

**Comment by nickcsbr** [comment]
> Engine Note still sounds incorrect compared to Real Hardware. Almost as if the pitch of it is too high?
> 
> 
> 
> I can try and capture some comparison Audio of both the Core and a Japanese Model 1 Mega Drive to demonstrate?

Yes, please. It is interesting that software emulators do this well while FPGA implementations (mister and Analogue) don't.

**Comment by birdybro** [comment]
@jotego - jt89 is a reimplementation of the SN76489AN, correct? The Sega Genesis, Master System, and Game Gear all used the slightly different SN76489A clones instead which has slightly different pseudo-randomness and the periodic noise is 16 stages instead of 15 stages.

Any chance a port could be added upstream into jt89_noise which switches the behavior between an SN76489A and an SN76489AN?

https://web.archive.org/web/20060517025848/http://www.smspower.org/maxim/docs/SN76489.txt - See section `5.2.1 The Linear Feedback Shift Register`.

Currently jt89_noise uses the 16 bit array, whereas the Genesis, SMS, and GG all used a clone which had a 15 bit array. The tapped bits are also different, so this could cause the perceived inaccuracies at the margins, like in this example.

**Comment by jotego** [comment]
Thank you, @birdybro, for the detailed description. However, the behavior for Genesis/SMS/Game Gear is what's already implemented, with a 16-bit shift register. Adding options for SG-1000 and Tandy 1000 is not connected to this issue report.

I have reviewed the noise channel code. This was already improved way back. Check this [commit](https://github.com/jotego/jt89/commit/13da1edcd56b15f3fe01d075acdc7ef48e879d2a). I have also read again the documentation you linked to, and everything seems fine.

We can try a couple of things, if you are willing to compile the core and test it for us. Please compile it with this [branch of jt89](https://github.com/jotego/jt89/tree/Genesis_MiSTer%23173).  I have made the noise link to channel 2 like the original hardware implementation. Let me know if it makes a difference.

**Comment by birdybro** [comment]
Thanks! I'll test that out!

**Comment by birdybro** [comment]
Oh before I do, it looks as though the update to jt89 in the Genesis core not too long ago already ported these changes:

https://github.com/MiSTer-devel/Genesis_MiSTer/commit/6d671f2951e44b638aa3917ab458ad89fc68d248

**Comment by birdybro** [comment]
Ugh... I forgot that I had already determined this is likely in ym2612 not in the sn76489... Sorry, got too excited.

**Comment by nickcsbr** [comment]
@birdybro which version of the core are you currently using? I am not sure, it's been a while since I last tested it. I'll try it tonight.

**Comment by nickcsbr** [comment]
I did try version v230224 and the problem is still there. For the world championship mode, selecting manual shift 7-speed, for the first GP I did the qualify lap with the engine sounding fine but when the race started, the sound is different already.

**Comment by birdybro** [comment]
Can you copy the rom to the s32x folder on the mister, rename the file extension to .32x, load it in the sega 32x core and see if it sounds better? There may be other bugs visually and stuff, but if it works and it sounds correct, then that means it's a bus arbitration timing issue.

**Comment by nickcsbr** [comment]
@birdybro I have tried your suggestion, it does work but the engine sound is still too erratic.

**Comment by birdybro** [comment]
I believe this engine sound is fed through the DAC in the ym2612 given the vgm logs above, but I'm not too sure. Sorry I'm unable to help with this.

**Comment by jotego** [comment]
Could someone compile the core muting PCM/FM/PSG to determine the specific sound source for the car engine?

**Comment by birdybro** [comment]
[GenesisMutedPSG.zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/12114289/GenesisMutedPSG.zip)

Build without PSG (I think). Just tied jt89's `sound` port to `0`.

**Comment by davedoc1984** [comment]
Played the game using Genesis BlastEm and the Engine Sound is much closer if not the same as it is on Real Hardware.

**Comment by birdybro** [comment]
Yes, BlastEm is the current leading emulator in accuracy of sound playback for the most part. Thanks for testing.

**Comment by davedoc1984** [comment]
To add further, when I play this in PAL 50hz Mode (EU Region) within the Core... the Engine sounds are much closer to the Real Hardware. In fact I can barely tell a difference. 

However, If I then boot the game with the 60hz Mode (US or JP Region) the Engine sounds are off. Like they are a higher pitch / frequency.

**Comment by birdybro** [comment]
Does the pitch sound better when you change the rom storage to ddr3 and reload the rom?

**Comment by davedoc1984** [comment]
(NTSC) Sounds just the same as before when using DDR3 and reloading the ROM.

**Comment by SegaSnatcher** [comment]
Curious how Nukeykt's core will handle this game.

**Comment by nickcsbr** [comment]
> Curious how Nukeykt's core will handle this game. 

Other software emulators handle this particular scenario quite well.

**Comment by nickcsbr** [comment]
Just an update, I have tried this game with the new Mega Drive core and the audio works exactly as expected.
